pgjdbc/src/main/java/org/postgresql/largeobject/BlobOutputStream.java

```diff
@@ -55,22 +59,37 @@ public BlobOutputStream(LargeObject lo) {
    */
   public BlobOutputStream(LargeObject lo, int bufferSize) {
     this.lo = lo;
-    this.bufferSize = bufferSize;
+    // Avoid "0" buffer size, and ensure the bufferSize will always be a power of two
+    this.bufferSize = Integer.highestOneBit(Math.max(bufferSize, 1));
+  }
+
+  /**
+   * Grows an internal buffer to ensure the extra bytes fit in the buffer.
+   * @param extraBytes the number of extra bytes that should fit in the buffer
+   * @return new buffer
+   */
+  private byte[] growBuffer(int extraBytes) {
+    byte[] buf = this.buf;
+    if (buf != null && (buf.length == bufferSize || buf.length - bufferPosition >= extraBytes)) {
+      // Buffer is already large enough
+      return buf;
+    }
+    // We use power-of-two buffers, so they align nicely with PostgreSQL's LargeObject slicing
+    // By default PostgreSQL slices the data in 2KiB chunks
+    int newSize = Math.min(bufferSize, Integer.highestOneBit(bufferPosition + extraBytes) * 2);
+    this.buf = buf = buf == null ? new byte[newSize] : Arrays.copyOf(buf, newSize);
```

Comment:
takes me more than a few seconds to grok what is happening here. comment?

Reply:
There was an issue though: `Arrays.copyOf(buf, newSize);` copies full contents of the buffer, however, we do not always have data in the buffer.
So I replaced it with explicit "buffer creating" and "copying the old data"